File: content/problems/1172-dinner-plate-stacks/solutions/ordered-nonfull-index-set/Solution.java
class DinnerPlates {
    private final int capacity;
    private final List<ArrayDeque<Integer>> stacks=new ArrayList<>();
    private final TreeSet<Integer> available=new TreeSet<>();
    public DinnerPlates(int capacity){
        this.capacity=capacity;
    }
    public void push(int val){
        int index;
        if(available.isEmpty()){
            index=stacks.size();
            stacks.add(new ArrayDeque<>());
        }else index=available.first();
        ArrayDeque<Integer> stack=stacks.get(index);
        stack.push(val);
        if(stack.size()==capacity)available.remove(index);
        else available.add(index);
    }
    public int pop(){
        return stacks.isEmpty()?-1:popAtStack(stacks.size()-1);
    }
    public int popAtStack(int index){
        if(index<0||index>=stacks.size()||stacks.get(index).isEmpty())return -1;
        int value=stacks.get(index).pop();
        available.add(index);
        while(!stacks.isEmpty()&&stacks.get(stacks.size()-1).isEmpty()){
            int last=stacks.size()-1;
            stacks.remove(last);
            available.remove(last);
        }
        return value;
    }
}

File: content/problems/1172-dinner-plate-stacks/solutions/ordered-nonfull-index-set/solution.cpp
class DinnerPlates {
    int capacity;
    vector<vector<int>> stacks;
    set<int> available;
    public:DinnerPlates(int capacity):capacity(capacity){
    }
    void push(int val){
        int index;
        if(available.empty()){
            index=stacks.size();
            stacks.push_back({
            });
        }else index=*available.begin();
        stacks[index].push_back(val);
        if((int)stacks[index].size()==capacity)available.erase(index);
        else available.insert(index);
    }
    int pop(){
        return stacks.empty()?-1:popAtStack(stacks.size()-1);
    }
    int popAtStack(int index){
        if(index<0||index>=(int)stacks.size()||stacks[index].empty())return -1;
        int value=stacks[index].back();
        stacks[index].pop_back();
        available.insert(index);
        while(!stacks.empty()&&stacks.back().empty()){
            int last=stacks.size()-1;
            stacks.pop_back();
            available.erase(last);
        }
        return value;
    }
};
